### PR TITLE
*: fix updating the column value when the column is dropping and in WriteOnly state

### LINKED ISSUE #15386: unexpected result when executing DDL  and DML concurrently
CODE master

```sql
create table t (a int, b int)
insert into t values(1,1);
```

connection 1

```sql
alter table t drop column b;
```
connection2 

```sql
update t  set a=2,b=2;
update t set a=10,where b=2;
select * from t;
```

result
mysql> Select * from t;
+------+
| a    |
+------+
|    2 |
+------+
1 row in set (0.00 sec)

### PR DESCRIPTION
**Pull request by zimulala**
### What problem does this PR solve?

Issue Number: close #15386 
```
create table t (a int, b int)
insert into t values(1,1);
```

Do the following:
| time  | connection1  |        connection2 |
|---|:---:|:---:|
| t1      |  drop column |
|t2     |   column state (write only) |  update  t set a =5 ,b=2|
|t3       | column state (write only) | update t set a=10 where b=2| 
|t4       | column state (none)    |      select * from t |

The result:
```
tidb> select * from t;
+------+
| a |
+------+
| 5 |
+------+
1 row in set (0.00 sec)
```

Problem Summary:
When the column is dropping and in WriteOnly state, the value of this column cannot be explicitly updated.

### What is changed and how it works?

What's Changed:
When the column is in a falling state and in WriteOnly state, if we update this column value explicitly, an error will be returned.

### Related changes

- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test

Side effects

- Performance regression
    - Consumes more CPU
    - Consumes more MEM
- Breaking backward compatibility

### CODE REVIEW

**Comment by zimulala** [comment]
PTAL @bb7133 @winoros @crazycs520

File: planner/core/logical_plan_builder.go [hunk 1]
```diff
@@ -3223,9 +3223,42 @@ func (b *PlanBuilder) buildUpdate(ctx context.Context, update *ast.UpdateStmt) (
 		tblID2table[id], _ = b.is.TableByID(id)
 	}
 	updt.TblColPosInfos, err = buildColumns2Handle(updt.OutputNames(), tblID2Handle, tblID2table, true)
+	if err == nil {
```
Comment (bb7133) [inline planner/core/logical_plan_builder.go]:
```suggestion
	if err != nil {
```
Better to use this style.

Comment (crazycs520) [reply planner/core/logical_plan_builder.go]: No, this logic is wrong

File: ddl/db_change_test.go [hunk 2]
```diff
@@ -542,17 +542,37 @@ func (s *testStateChangeSuite) TestDeleteOnly(c *C) {
...
+	_, err = s.se.Execute(context.Background(), "insert into tt (c1, c4) values(8, 8)")
+	c.Assert(err, IsNil)
+	defer s.se.Execute(context.Background(), "drop table tt")
+
+	sqls := make([]sqlWithErr, 2)
+	sqls[0] = sqlWithErr{"update t set c1='5', c3='2020-03-01';", errors.New("[planner:1054]Unknown column 'c3' in 'field list'")}
+	sqls[1] = sqlWithErr{"update t t1, tt t2 set t1.c1='5', t1.c3='2020-03-01', t2.c1='10' where t1.c4=t2.c4",
+		errors.New("[planner:1054]Unknown column 'c3' in 'field list'")}
+	// TODO: Fix the case of sqls[2].
+	// sqls[2] = sqlWithErr{"update t set c1='5' where c3='2017-07-01';", errors.New("[planner:1054]Unknown column 'c3' in 'field list'")}
```
Comment (tangenta) [inline ddl/db_change_test.go]: Do we need another issue to track this problem?

Comment (zimulala) [reply ddl/db_change_test.go]: Done. #15567

**Comment by zimulala** [comment]
PTAL @bb7133 @tangenta @winoros

File: planner/core/logical_plan_builder.go [hunk 3]
```diff
@@ -3226,9 +3226,42 @@ func (b *PlanBuilder) buildUpdate(ctx context.Context, update *ast.UpdateStmt) (
...
+		err = checkUpdateList(b.ctx, tblID2table, updt)
```
Comment (lysu) [inline planner/core/logical_plan_builder.go:3230]: it seems we only check it in update above normal executor..so point above point-get maybe hit this bug? :sob:

Comment (zimulala) [reply planner/core/logical_plan_builder.go:3230]: @lysu  mmm... Then I should fix this or you should fix it?

Comment (lysu) [reply planner/core/logical_plan_builder.go:3230]: let me see see- - 

I'd fixed in https://github.com/pingcap/tidb/pull/23491 during a refactor, but maybe it's not suitable to cherry-pick to 4.0 or 3.0

**Pull request merged**